src/Internal/PsrRequestStream.php

```diff
@@ -0,0 +1,192 @@
+
+use Amp\ByteStream\InputStream;
+use Psr\Http\Message\StreamInterface;
+
+use function Amp\Promise\timeout;
+use function Amp\Promise\wait;
+
+/**
+ * @internal
+ */
+final class PsrRequestStream implements StreamInterface
+{
+    private const DEFAULT_TIMEOUT = 5000;
+
+    /**
+     * @var InputStream
+     */
+    private $stream;
+
+    /**
+     * @var int
+     */
+    private $timeout;
+
+    /**
+     * @var string
+     */
+    private $buffer = '';
+
+    /**
+     * @var bool
+     */
+    private $isEof = false;
+
+    public function __construct(InputStream $stream, int $timeout = self::DEFAULT_TIMEOUT)
+    {
+        $this->stream = $stream;
+        $this->timeout = $timeout;
+    }
+
+    public function __toString()
+    {
+        try {
+            return $this->getContents();
+        } catch (\Throwable $e) {
+            return '';
+        }
+    }
+
+    public function close(): void
+    {
+        unset($this->stream);
+    }
+
+    public function eof(): bool
+    {
+        return $this->isEof;
+    }
+
+    public function tell()
+    {
+        throw new \RuntimeException("Source stream is not seekable");
+    }
+
+    /**
+     * @return null
+     */
+    public function getSize()
+    {
+        return null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isSeekable()
+    {
+        return false;
+    }
+
+    /**
+     * @param int $offset
+     * @param int $whence
+     * @return void
+     */
+    public function seek($offset, $whence = SEEK_SET)
+    {
+        throw new \RuntimeException("Source stream is not seekable");
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind()
+    {
+        throw new \RuntimeException("Source stream is not seekable");
+    }
+
+    /**
+     * @return bool
+     */
+    public function isWritable()
+    {
+        return false;
+    }
+
+    public function write($string)
+    {
+        throw new \RuntimeException("Source stream is not writable");
+    }
+
+    /**
+     * @param string|null $key
+     * @return array|null
+     */
+    public function getMetadata($key = null)
+    {
+        return isset($key) ? null : [];
+    }
+
+    /**
+     * @return null
+     */
+    public function detach()
+    {
+        unset($this->stream);
+
+        return null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isReadable()
+    {
+        return isset($this->stream);
+    }
+
+    /**
+     * @param int $length
+     * @return string
+     */
+    public function read($length)
+    {
+        while (!$this->isEof && \strlen($this->buffer) < $length) {
+            $this->buffer .= $this->readFromStream();
+        }
+
+        $data = \substr($this->buffer, 0, $length);
+        $this->buffer = \substr($this->buffer, \strlen($data));
+
+        return $data;
+    }
+
+    private function readFromStream(): string
+    {
+        $data = wait(timeout($this->getOpenStream()->read(), $this->timeout));
```

Comment:
We can't use `wait`, because it has issues with nesting, see https://github.com/amphp/amp/pull/311 and https://github.com/amphp/amp/pull/313. This means we need to buffer the entire body for these.

Reply:
Hmm, that's really bad: the body can be very big in some cases. Is there any workaround that will allow as to read the body by parts? Maybe some non-promise generator?

Reply:
The only one would be writing huge responses to files, I guess.

Reply:
Is there any chance of solving these issues in the future? In this case I can buffer the entire body for now, but keep the infrastructure to restore current behavior in future.

Is there any (relatively) simple way to reproduce the issues using my code? Maybe I'll think something out.

Reply:
Yes, it's very easy to reproduce, just call that code inside Loop::run() and it'll stop the outer Loop::run(), too.